rgbd/src/server_rgbd.cpp

```diff
@@ -0,0 +1,127 @@
+#include "rgbd/server_rgbd.h"
+
+#include <opencv2/core/mat.hpp>
+#include <opencv2/imgcodecs.hpp>
+
+#include <rgbd_msgs/RGBD.h>
+
+#include <ros/node_handle.h>
+
+#include <tue/serialization/conversions.h>
+
+#include "rgbd/image.h"
+#include "rgbd/serialization.h"
+
+namespace rgbd {
+
+const int ServerRGBD::MESSAGE_VERSION = 2;
+
+// ----------------------------------------------------------------------------------------
+
+ServerRGBD::ServerRGBD()
+{
+}
+
+// ----------------------------------------------------------------------------------------
+
+ServerRGBD::~ServerRGBD()
+{
+    nh_.shutdown();
+    send_thread_.join();
+    service_thread_.join();
+}
+
+// ----------------------------------------------------------------------------------------
+
+void ServerRGBD::initialize(const std::string& name, RGBStorageType rgb_type, DepthStorageType depth_type, const float service_freq)
+{
+    pub_image_ = nh_.advertise<rgbd_msgs::RGBD>(name, 1);
+    rgb_type_ = rgb_type;
+    depth_type_ = depth_type;
+
+    nh_.setCallbackQueue(&cb_queue_);
+    service_server_ = nh_.advertiseService(name, &ServerRGBD::serviceCallback, this);
+    service_thread_ = boost::thread(&ServerRGBD::serviceThreadFunc, this, service_freq);
+}
+
+// ----------------------------------------------------------------------------------------
+
+void ServerRGBD::send(const Image& image, bool threaded)
+{
+    threaded = false; // Running threaded causes a memory leak.
+    if (!threaded)
+    {
+        sendImpl(image);
+        return;
+    }
+
+    send_thread_ = boost::thread(&ServerRGBD::sendImpl, this, image);
+}
+
+// ----------------------------------------------------------------------------------------
+
+void ServerRGBD::sendImpl(const Image& image)
+{
+    {
+        boost::unique_lock<boost::mutex> ul(image_mutex_);
+        image_ = image.clone();
+    }
+
+    if (pub_image_.getNumSubscribers() == 0)
+        return;
+
+    rgbd_msgs::RGBD msg;
+    msg.version = MESSAGE_VERSION;
+
+    std::stringstream stream;
+    tue::serialization::OutputArchive a(stream);
+    serialize(image, a, rgb_type_, depth_type_);
+    tue::serialization::convert(stream, msg.rgb);
+
+    pub_image_.publish(msg);
+}
+
+bool ServerRGBD::serviceCallback(rgbd_msgs::GetRGBDRequest& req, rgbd_msgs::GetRGBDResponse& resp)
+{
+    rgbd::Image image;
+    {
+        boost::unique_lock<boost::mutex> ul(image_mutex_);
```

Comment:
Maybe out of scope, but I'm pretty sure this can be improved so that you don't need mutexes.

Reply:
E.g., https://thispointer.com/c11-multithreading-part-8-stdfuture-stdpromise-and-returning-values-from-thread/

Reply:
I haven't been able to find info about. The reference returned by `future::get` staying valid after calling `promise::set_value()`. As this is unclear, I recommend to stick to current implementation.